### PR TITLE
Fix memory leak in WebSocket and refactor code

### PR DESCRIPTION
**Pull request by riyafa**
## Automation tests
 - Unit tests 
   > N/A
 - Integration tests
   > N/A

## Security checks
 N/A

### CODE REVIEW

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java [hunk 1]
```diff
@@ -40,11 +48,39 @@ public static WebSocketSessionImpl getSession(ChannelHandlerContext ctx,
...
+    public static WebSocketMessageImpl getWebSocketMessage(BinaryWebSocketFrame binaryWebSocketFrame) {
+        ByteBuf byteBuf = binaryWebSocketFrame.content();
+        boolean finalFragment = binaryWebSocketFrame.isFinalFragment();
+        ByteBuffer byteBuffer = getClonedByteBuffer(byteBuf);
+        WebSocketMessageImpl webSocketBinaryMessage = new WebSocketBinaryMessageImpl(byteBuffer, finalFragment);
+        binaryWebSocketFrame.release();
+        return webSocketBinaryMessage;
+    }
+
+    public static ByteBuffer getClonedByteBuffer(ByteBuf buf) {
```
Comment (anupama-pathirage) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java]: This method can be private

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java [hunk 2]
```diff
@@ -40,11 +48,39 @@ public static WebSocketSessionImpl getSession(ChannelHandlerContext ctx,
...
+        boolean isFinalFragment = textWebSocketFrame.isFinalFragment();
+        WebSocketMessageImpl webSocketTextMessage = new WebSocketTextMessageImpl(text, isFinalFragment);
+        textWebSocketFrame.release();
+        return webSocketTextMessage;
+    }
+
+    public static WebSocketMessageImpl getWebSocketMessage(BinaryWebSocketFrame binaryWebSocketFrame) {
+        ByteBuf byteBuf = binaryWebSocketFrame.content();
+        boolean finalFragment = binaryWebSocketFrame.isFinalFragment();
+        ByteBuffer byteBuffer = getClonedByteBuffer(byteBuf);
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java]: Can we swap above two lines. Second breaks the vertical affinity.

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java [hunk 3]
```diff
@@ -40,11 +48,39 @@ public static WebSocketSessionImpl getSession(ChannelHandlerContext ctx,
...
 
-    public static String getSubProtocol(HttpRequest request) {
-        HttpHeaders headers = request.headers();
-        if (headers.contains(Constants.WEBSOCKET_HEADER_SUBPROTOCOL)) {
-            return headers.get(Constants.WEBSOCKET_HEADER_SUBPROTOCOL);
-        }
-        return null;
+    public static WebSocketControlMessage getWebsocketControlMessage(WebSocketFrame webSocketFrame,
+                                                                     WebSocketControlSignal controlSignal) {
+        ByteBuf byteBuf = webSocketFrame.content();
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java]: content

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java [hunk 4]
```diff
@@ -40,11 +48,39 @@ public static WebSocketSessionImpl getSession(ChannelHandlerContext ctx,
...
+        ByteBuffer byteBuffer = getClonedByteBuffer(byteBuf);
```
Comment (shafreenAnfar) [inline components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java]: And here clonedContent

Comment (riyafa) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java]: Do you mean that this method name be renamed?

File: components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java (hunk 3 above)
Comment (riyafa) [reply components/org.wso2.transport.http.netty/src/main/java/org/wso2/transport/http/netty/internal/websocket/WebSocketUtil.java]: Renamed method

**Pull request merged**